lib/internal/streams/end-of-stream.js

```diff
@@ -59,25 +59,23 @@ function eos(stream, opts, callback) {
     };
   }
 
-  let readable = opts.readable ||
+  const readable = opts.readable ||
     (opts.readable !== false && isReadable(stream));
-  let writable = opts.writable ||
+  const writable = opts.writable ||
     (opts.writable !== false && isWritable(stream));
 
   const onlegacyfinish = () => {
     if (!stream.writable) onfinish();
   };
 
   const onfinish = () => {
-    writable = false;
     writableFinished = true;
-    if (!readable) callback.call(stream);
+    if (!readable || readableEnded) callback.call(stream);
```

Comment:
This seems a significant change that merit a unit test.

Reply:
https://github.com/nodejs/node/blob/master/test/parallel/test-stream-finished.js#L207